File: leetcode/binary-search.py
# https://leetcode.com/problems/binary-search/description/


class Solution:
    def search(self, nums: list[int], target: int) -> int:
        lo: int = 0
        hi: int = len(nums) - 1

        while lo <= hi:
            mid: int = int((lo + hi) / 2)
            if nums[mid] < target:
                lo = mid + 1
            elif nums[mid] > target:
                hi = mid - 1
            else:
                return mid

        return -1

File: leetcode/binary-search.java
//https://leetcode.com/problems/binary-search/

class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length-1;
        
        while(left <= right){
            // Emulated 'mid' pointer for slight space efficiency increase
            if(nums[(left+right)/2] == target) return (left+right)/2;
            else if(nums[(left+right)/2] > target) right = ((left+right)/2)-1;
            else left = ((left+right)/2)+1;
        }
        
        return -1;
    }
}